Exception in thread "main" java.util.concurrent.ExecutionException: java.lang.NumberFormatException: For input string: "20160909030348"

While running the `latest` docker image over my project's requirements.txt, came across a traceback. I traced the failure down to `graphql-core==3.1.3` in the requirements file. Turns out this is reproducible with just `docker run -it --rm pilosus/pip-license-checker java -jar app.jar graphql-core` too.

#### Traceback

```clojure
Exception in thread "main" java.util.concurrent.ExecutionException: java.lang.NumberFormatException: For input string: "20160909030348"
	at java.base/java.util.concurrent.FutureTask.report(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.get(Unknown Source)
	at clojure.core$deref_future.invokeStatic(core.clj:2300)
	at clojure.core$future_call$reify__8454.deref(core.clj:6974)
	at clojure.core$deref.invokeStatic(core.clj:2320)
	at clojure.core$deref.invoke(core.clj:2306)
	at clojure.core$map$fn__5866.invoke(core.clj:2753)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:58)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5402.invokeStatic(core.clj:137)
	at clojure.core$concat$fn__5493.invoke(core.clj:725)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5402.invokeStatic(core.clj:137)
	at pip_license_checker.core$process_requirements.invokeStatic(core.clj:80)
	at pip_license_checker.core$_main.invokeStatic(core.clj:222)
	at pip_license_checker.core$_main.doInvoke(core.clj:217)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at pip_license_checker.core.main(Unknown Source)
Caused by: java.lang.NumberFormatException: For input string: "20160909030348"
	at java.base/java.lang.NumberFormatException.forInputString(Unknown Source)
	at java.base/java.lang.Integer.parseInt(Unknown Source)
	at java.base/java.lang.Integer.parseInt(Unknown Source)
	at pip_license_checker.version$parse_number.invokeStatic(version.clj:35)
	at pip_license_checker.version$parse_letter_version.invokeStatic(version.clj:50)
	at pip_license_checker.version$validate_version.invokeStatic(version.clj:98)
	at pip_license_checker.version$parse_version.invokeStatic(version.clj:125)
	at pip_license_checker.pypi$get_releases$fn__4439.invoke(pypi.clj:58)
	at clojure.core$map$fn__5866.invoke(core.clj:2755)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5402.invokeStatic(core.clj:137)
	at clojure.core$filter$fn__5893.invoke(core.clj:2809)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5402.invokeStatic(core.clj:137)
	at clojure.core$filter$fn__5893.invoke(core.clj:2809)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5402.invokeStatic(core.clj:137)
	at clojure.core$filter$fn__5893.invoke(core.clj:2809)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5402.invokeStatic(core.clj:137)
	at pip_license_checker.version$filter_versions.invokeStatic(version.clj:436)
	at pip_license_checker.version$get_version.invokeStatic(version.clj:463)
	at pip_license_checker.pypi$get_requirement_version.invokeStatic(pypi.clj:72)
	at pip_license_checker.pypi$requirement__GT_license.invokeStatic(pypi.clj:155)
	at pip_license_checker.pypi$get_parsed_requiements$fn__4479.invoke(pypi.clj:188)
	at clojure.core$pmap$fn__8462$fn__8463.invoke(core.clj:7022)
	at clojure.core$binding_conveyor_fn$fn__5754.invoke(core.clj:2030)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,10 @@ This change log follows the conventions of [keepachangelog.com](http://keepachan
 
 ## [Unreleased]
 
+## [0.33.0] - 2022-06-25
+### Fixed
+- Parsing long numbers in Python versions ((#99)[https://github.com/pilosus/pip-license-checker/issues/99])
+
 ## [0.32.0] - 2022-06-23
 ### Added
 - `lein ancient` to check outdated deps
@@ -267,7 +271,8 @@ weak copyleft types.
 ### Added
 - Structure for Leiningen app project
 
-[Unreleased]: https://github.com/pilosus/pip-license-checker/compare/0.32.0...HEAD
+[Unreleased]: https://github.com/pilosus/pip-license-checker/compare/0.33.0...HEAD
+[0.33.0]: https://github.com/pilosus/pip-license-checker/compare/0.32.0...0.33.0
 [0.32.0]: https://github.com/pilosus/pip-license-checker/compare/0.31.0...0.32.0
 [0.31.0]: https://github.com/pilosus/pip-license-checker/compare/0.30.0...0.31.0
 [0.30.0]: https://github.com/pilosus/pip-license-checker/compare/0.29.0...0.30.0

--- a/project.clj
+++ b/project.clj
@@ -1,4 +1,4 @@
-(defproject org.clojars.vrs/pip-license-checker "0.32.0"
+(defproject org.clojars.vrs/pip-license-checker "0.33.0-SNAPSHOT"
   :description "License compliance tool to identify dependencies license names and types: permissive, copyleft, proprietory, etc."
   :url "https://github.com/pilosus/pip-license-checker"
   :license {:name "Eclipse Public License 2.0 OR GNU GPL v2+ with Classpath exception"

--- a/src/pip_license_checker/version.clj
+++ b/src/pip_license_checker/version.clj
@@ -24,6 +24,7 @@
 
 ;; Parse version
 
+(def regex-number #"^\d+$")
 (def regex-split-comma #",")
 (def regex-specifier #"(?<op>(===|==|~=|!=|>=|<=|<|>))(?<version>(.*))")
 
@@ -32,7 +33,11 @@
 (defn parse-number
   "Parse number string into integer or return 0"
   [number]
-  (if (not number) 0 (Integer/parseInt number)))
+  (if (or (not number)
+          ;; just to make sure no code execution is possible with read-string
+          (not (re-find #"^-?\d+\.?\d*$" number)))
+    0
+    (read-string number)))
 
 (s/fdef parse-letter-version
   :args (s/cat :letter ::sp/matched-version-part

--- a/test/pip_license_checker/version_test.clj
+++ b/test/pip_license_checker/version_test.clj
@@ -112,7 +112,19 @@
      :post ["post" 345]
      :dev ["dev" 456]
      :local nil}
-    "Release, pre, post and dev version"]])
+    "Release, pre, post and dev version"]
+   ["1.0.dev20160909030348"
+    {:orig "1.0.dev20160909030348"
+     :epoch 0
+     :release [1 0]
+     :pre nil
+     :post nil
+     :dev ["dev" 20160909030348]
+     :local nil}
+    "Make sure long numbers handled correctly"]
+   ["1.0.dev(exploit-run)"
+    nil
+    "Make sure no read-string expoit is possible"]])
 
 (deftest test-parse-version
   (testing "Version parsing"

